File: backend-spring/src/main/java/com/ivyts/backend/service/EnrollmentService.java
package com.ivyts.backend.service;

import com.ivyts.backend.common.exception.ApiException;
import com.ivyts.backend.domain.course.Course;
import com.ivyts.backend.domain.enrollment.Enrollment;
import com.ivyts.backend.domain.enrollment.EnrollmentStatus;
import com.ivyts.backend.domain.enrollment.LessonProgressItem;
import com.ivyts.backend.domain.lesson.Lesson;
import com.ivyts.backend.domain.user.User;
import com.ivyts.backend.domain.user.UserRole;
import com.ivyts.backend.notification.NotificationEventsService;
import com.ivyts.backend.security.AuthUser;
import com.ivyts.backend.service.coursestore.CourseStore;
import com.ivyts.backend.service.coursestore.LessonStore;
import com.ivyts.backend.service.enrollmentstore.EnrollmentStore;
import com.ivyts.backend.service.userstore.UserStore;
import com.ivyts.backend.web.enrollment.EnrollmentMapper;
import com.ivyts.backend.web.enrollment.dto.CreateEnrollmentRequest;
import com.ivyts.backend.web.enrollment.dto.UpdateProgressRequest;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

@Service
public class EnrollmentService {

    private final EnrollmentStore enrollmentStore;
    private final CourseStore courseStore;
    private final LessonStore lessonStore;
    private final UserStore userStore;
    private final EnrollmentMapper enrollmentMapper;
    private final NotificationEventsService notificationEventsService;

    public EnrollmentService(
        EnrollmentStore enrollmentStore,
        CourseStore courseStore,
        LessonStore lessonStore,
        UserStore userStore,
        EnrollmentMapper enrollmentMapper,
        NotificationEventsService notificationEventsService
    ) {
        this.enrollmentStore = enrollmentStore;
        this.courseStore = courseStore;
        this.lessonStore = lessonStore;
        this.userStore = userStore;
        this.enrollmentMapper = enrollmentMapper;
        this.notificationEventsService = notificationEventsService;
    }

    public Map<String, Object> enrollCourse(CreateEnrollmentRequest request, AuthUser authUser) {
        ensureStudent(authUser);
        Course course = findPublishedCourseOrThrow(request.courseId());

        enrollmentStore.findByCourseAndStudent(course.getId(), authUser.userId())
            .ifPresent(existing -> {
                throw new ApiException(HttpStatus.CONFLICT, "You are already enrolled in this course");
            });

        Enrollment enrollment = new Enrollment();
        enrollment.setStudent(authUser.userId());
        enrollment.setCourse(course.getId());
        enrollment.setStatus(EnrollmentStatus.ACTIVE);
        enrollment.setProgressPercent(0);
        enrollment.setCompletedLessonIds(new ArrayList<>());
        enrollment.setLessonProgress(new ArrayList<>());
        enrollment.setEnrolledAt(Instant.now());
        enrollment = enrollmentStore.save(enrollment);
        User student = userStore.findById(authUser.userId())
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Student not found"));
        User owner = userStore.findById(course.getOwner())
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Course owner not found"));
        notificationEventsService.emitEnrollmentCreated(
            student.getId(),
            student.getFullName(),
            course.getId(),
            course.getTitle(),
            owner.getId(),
            owner.getRole().name().toLowerCase()
        );

        return toEnrollmentView(enrollment);
    }

    public List<Map<String, Object>> getMyEnrollments(AuthUser authUser) {
        ensureStudent(authUser);
        return enrollmentStore.findByStudentAndStatusInOrderByCreatedAtDesc(
                authUser.userId(),
                List.of(EnrollmentStatus.ACTIVE, EnrollmentStatus.COMPLETED)
            )
            .stream()
            .map(this::toEnrollmentView)
            .toList();
    }

    public List<Map<String, Object>> getCourseEnrollments(String courseId, AuthUser authUser) {
        Course course = findCourseOrThrow(courseId);
        if (!canManageCourse(authUser, course)) {
            throw new ApiException(HttpStatus.FORBIDDEN, "You do not have permission to view these enrollments");
        }

        return enrollmentStore.findByCourseOrderByCreatedAtDesc(courseId).stream()
            .map(this::toEnrollmentView)
            .toList();
    }

    public Map<String, Object> updateLearningProgress(String courseId, UpdateProgressRequest request, AuthUser authUser) {
        ensureStudent(authUser);
        Enrollment enrollment = findEnrollmentOrThrow(courseId, authUser.userId());
        Lesson lesson = lessonStore.findById(request.lessonId())
            .filter(current -> current.getCourse().equals(courseId))
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Lesson not found in this course"));

        int totalLessons = lessonStore.findByCourseOrderByOrderAsc(courseId).size();
        Instant nextAccessedAt = request.lastAccessedAt() != null ? request.lastAccessedAt() : Instant.now();

        LessonProgressItem progressItem = enrollment.getLessonProgress().stream()
            .filter(item -> item.getLesson().equals(request.lessonId()))
            .findFirst()
            .orElse(null);

        if (progressItem == null) {
            progressItem = new LessonProgressItem();
            progressItem.setLesson(request.lessonId());
            progressItem.setWatchedSeconds(0);
            progressItem.setCompleted(false);
            enrollment.getLessonProgress().add(progressItem);
        }

        if (request.watchedSeconds() != null) {
            progressItem.setWatchedSeconds(Math.max(progressItem.getWatchedSeconds(), request.watchedSeconds()));
        }
        if (request.isCompleted() != null) {
            progressItem.setCompleted(request.isCompleted());
        }
        progressItem.setLastAccessedAt(nextAccessedAt);
        progressItem.setCompletedAt(progressItem.isCompleted()
            ? progressItem.getCompletedAt() != null ? progressItem.getCompletedAt() : Instant.now()
            : null);

        List<String> completedLessonIds = enrollment.getLessonProgress().stream()
            .filter(LessonProgressItem::isCompleted)
            .map(LessonProgressItem::getLesson)
            .distinct()
            .toList();

        enrollment.setCompletedLessonIds(new ArrayList<>(completedLessonIds));
        enrollment.setProgressPercent(computeProgressPercent(completedLessonIds.size(), totalLessons));
        enrollment.setLastLessonId(lesson.getId());
        enrollment.setStartedAt(enrollment.getStartedAt() != null ? enrollment.getStartedAt() : Instant.now());
        enrollment.setStatus(enrollment.getProgressPercent() == 100 && totalLessons > 0 ? EnrollmentStatus.COMPLETED : EnrollmentStatus.ACTIVE);
        enrollment.setCompletedAt(enrollment.getStatus() == EnrollmentStatus.COMPLETED
            ? enrollment.getCompletedAt() != null ? enrollment.getCompletedAt() : Instant.now()
            : null);

        enrollment = enrollmentStore.save(enrollment);
        return toEnrollmentView(enrollment);
    }

    private Course findPublishedCourseOrThrow(String courseId) {
        Course course = findCourseOrThrow(courseId);
        if (!course.isPublished()) {
            throw new ApiException(HttpStatus.NOT_FOUND, "Course not found");
        }
        return course;
    }

    private Course findCourseOrThrow(String courseId) {
        return courseStore.findById(courseId)
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Course not found"));
    }

    private Enrollment findEnrollmentOrThrow(String courseId, String studentId) {
        return enrollmentStore.findByCourseAndStudent(courseId, studentId)
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Enrollment not found"));
    }

    private Map<String, Object> toEnrollmentView(Enrollment enrollment) {
        User student = userStore.findById(enrollment.getStudent())
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Student not found"));
        Course course = findCourseOrThrow(enrollment.getCourse());
        return enrollmentMapper.toEnrollmentView(enrollment, student, course);
    }

    private int computeProgressPercent(int completedCount, int totalLessons) {
        if (totalLessons <= 0) {
            return 0;
        }
        return Math.round((completedCount * 100.0f) / totalLessons);
    }

    private boolean canManageCourse(AuthUser authUser, Course course) {
        return authUser.role() == UserRole.ADMIN
            || (authUser.role() == UserRole.TEACHER && authUser.userId().equals(course.getOwner()));
    }

    private void ensureStudent(AuthUser authUser) {
        if (authUser.role() != UserRole.STUDENT) {
            throw new ApiException(HttpStatus.FORBIDDEN, "Only students can access this resource");
        }
    }
}

File: archive/backend-node/src/services/enrollment.service.ts
// Enrollment lifecycle: join a course, fetch student roster and persist lesson-level progress.
import { Course, Enrollment, Lesson } from '../models/index.js';
import { AuthUser } from '../types/auth.js';
import { HTTP_STATUS } from '../constants/http-status.js';
import { ApiError } from '../utils/api-error.js';
import { canManageCourse } from '../utils/course-permission.js';
import { mapEnrollment } from '../utils/enrollment-transform.js';
import { CreateEnrollmentInput, UpdateProgressInput } from '../validations/enrollment.validation.js';
import { emitEnrollmentCreated } from './notification-events.service.js';

async function findPublishedCourseOrThrow(courseId: string) {
  const course = await Course.findById(courseId).populate('owner', 'fullName email');

  if (!course || !course.isPublished) {
    throw new ApiError(HTTP_STATUS.NOT_FOUND, 'Course not found');
  }

  return course;
}

async function findEnrollmentOrThrow(courseId: string, studentId: string) {
  const enrollment = await Enrollment.findOne({
    course: courseId,
    student: studentId,
  })
    .populate('course', 'title slug thumbnail category level lessonCount totalDuration')
    .populate('student', 'fullName email');

  if (!enrollment) {
    throw new ApiError(HTTP_STATUS.NOT_FOUND, 'Enrollment not found');
  }

  return enrollment;
}

function computeProgressPercent(completedCount: number, totalLessons: number): number {
  if (totalLessons <= 0) {
    return 0;
  }

  return Math.round((completedCount / totalLessons) * 100);
}

export async function enrollCourse(
  input: CreateEnrollmentInput,
  user: AuthUser,
): Promise<ReturnType<typeof mapEnrollment>> {
  const course = await findPublishedCourseOrThrow(input.courseId);
  const existingEnrollment = await Enrollment.findOne({
    student: user.userId,
    course: input.courseId,
  });

  if (existingEnrollment) {
    throw new ApiError(HTTP_STATUS.CONFLICT, 'You are already enrolled in this course');
  }

  await Enrollment.create({
    student: user.userId,
    course: course._id,
    status: 'active',
    progressPercent: 0,
    completedLessonIds: [],
    lessonProgress: [],
    enrolledAt: new Date(),
  });

  const hydrated = await findEnrollmentOrThrow(course._id.toString(), user.userId);
  const student = hydrated.student as unknown;
  const courseOwner = course.owner as unknown;
  const ownerRole =
    courseOwner && typeof courseOwner === 'object' && 'role' in (courseOwner as Record<string, unknown>)
      ? String((courseOwner as { role?: unknown }).role ?? '') || undefined
      : undefined;
  emitEnrollmentCreated({
    studentUserId: user.userId,
    studentName:
      student && typeof student === 'object' && 'fullName' in (student as Record<string, unknown>)
        ? String((student as { fullName?: unknown }).fullName ?? user.email)
        : user.email,
    courseId: course._id.toString(),
    courseTitle: course.title,
    ownerId:
      courseOwner && typeof courseOwner === 'object' && '_id' in (courseOwner as Record<string, unknown>)
        ? String((courseOwner as { _id: unknown })._id)
        : String(course.owner),
    ownerRole: ownerRole === 'admin' || ownerRole === 'teacher' || ownerRole === 'student' ? ownerRole : undefined,
  });
  return mapEnrollment(hydrated);
}

export async function getMyEnrollments(user: AuthUser): Promise<Array<ReturnType<typeof mapEnrollment>>> {
  const enrollments = await Enrollment.find({
    student: user.userId,
    status: { $in: ['active', 'completed'] },
  })
    .populate('course', 'title slug thumbnail category level lessonCount totalDuration')
    .populate('student', 'fullName email')
    .sort({ createdAt: -1 });

  return enrollments.map((enrollment) => mapEnrollment(enrollment));
}

export async function getCourseEnrollments(
  courseId: string,
  user: AuthUser,
): Promise<Array<ReturnType<typeof mapEnrollment>>> {
  const course = await Course.findById(courseId).populate('owner', 'fullName email');

  if (!course) {
    throw new ApiError(HTTP_STATUS.NOT_FOUND, 'Course not found');
  }

  if (user.role !== 'admin' && !canManageCourse(user, course)) {
    throw new ApiError(HTTP_STATUS.FORBIDDEN, 'You do not have permission to view these enrollments');
  }

  const enrollments = await Enrollment.find({ course: courseId })
    .populate('course', 'title slug thumbnail category level lessonCount totalDuration')
    .populate('student', 'fullName email')
    .sort({ createdAt: -1 });

  return enrollments.map((enrollment) => mapEnrollment(enrollment));
}

export async function updateLearningProgress(
  courseId: string,
  input: UpdateProgressInput,
  user: AuthUser,
): Promise<ReturnType<typeof mapEnrollment>> {
  const enrollment = await findEnrollmentOrThrow(courseId, user.userId);
  const lesson = await Lesson.findOne({
    _id: input.lessonId,
    course: courseId,
  });

  if (!lesson) {
    throw new ApiError(HTTP_STATUS.NOT_FOUND, 'Lesson not found in this course');
  }

  const totalLessons = await Lesson.countDocuments({ course: courseId });
  const progressIndex = enrollment.lessonProgress.findIndex(
    (item) => item.lesson.toString() === input.lessonId,
  );
  const nextAccessedAt = input.lastAccessedAt ? new Date(input.lastAccessedAt) : new Date();

  if (progressIndex >= 0) {
    const current = enrollment.lessonProgress[progressIndex];
    current.watchedSeconds =
      typeof input.watchedSeconds === 'number'
        ? Math.max(current.watchedSeconds, input.watchedSeconds)
        : current.watchedSeconds;
    current.isCompleted = input.isCompleted ?? current.isCompleted;
    current.lastAccessedAt = nextAccessedAt;
    current.completedAt = current.isCompleted ? current.completedAt ?? new Date() : undefined;
  } else {
    enrollment.lessonProgress.push({
      lesson: lesson._id,
      watchedSeconds: input.watchedSeconds ?? 0,
      isCompleted: input.isCompleted ?? false,
      completedAt: input.isCompleted ? new Date() : undefined,
      lastAccessedAt: nextAccessedAt,
    });
  }

  const completedIds = enrollment.lessonProgress
    .filter((item) => item.isCompleted)
    .map((item) => item.lesson);

  enrollment.completedLessonIds = completedIds;
  enrollment.progressPercent = computeProgressPercent(completedIds.length, totalLessons);
  enrollment.lastLessonId = lesson._id;
  enrollment.startedAt = enrollment.startedAt ?? new Date();
  enrollment.status = enrollment.progressPercent === 100 && totalLessons > 0 ? 'completed' : 'active';
  enrollment.completedAt = enrollment.status === 'completed' ? enrollment.completedAt ?? new Date() : undefined;

  await enrollment.save();

  const updated = await findEnrollmentOrThrow(courseId, user.userId);
  return mapEnrollment(updated);
}
